### PR TITLE
Remove noisy documentation preview comment

### PR DESCRIPTION
**Pull request by tobiasdiez**
The comment linking to the built docs is noisy and doesn't add much value: the docs can also easily be accessed by clicking on the "Successfully deploy preview" status check or via the summary of the deploy workflow.

On the other hand, it always triggers a notification. Hence, we remove the comment.

### :memo: Checklist

- [ ] The title is concise, informative, and self-explanatory.
- [ ] The description explains in detail what this PR is about.
- [ ] I have linked a relevant issue or discussion.
- [ ] I have created tests covering the changes.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by mkoeppe** [comment]
-1 on the removal.
This notification is a useful feature. I find good value in it myself. And it's important on the scope of the project because it encourages looking at the changes to the formatted documentation when it is ready. This is a step that is too often skipped -- something that I know from making large-scale fixes of the RST markup across the Sage library.

People who do not want to see it can trivially filter it out in their email system.

**Comment by tobiasdiez** [comment]
> People who do not want to see it can trivially filter it out in their email system.

I don't get any emails from github. I and other are annoyed by the notifications in the github interface, which you cannot filter.

**Comment by mkoeppe** [comment]
> the functionality to preview the docs is still there of course, you just have to click in a different place:

Of course it's still there. The valuable point of the notification is ... to be notified, the importance of which is explained in detail in https://github.com/sagemath/sage/pull/37387#issuecomment-1951395012

**Comment by saraedum** [comment]
Could we drop the notification but change the PR checklist to tell people to check the generated notification?

**Comment by mantepse** [comment]
In the case there might be a vote at some point: I am strongly in favour of removing it.  I check the doc before I give Positive review, but the bot notifies me after every push, and inconveniently a few hours later.

**Comment by saraedum** [comment]
> In the case there might be a vote at some point: I am strongly in favour of removing it. I check the doc before I give Positive review, but the bot notifies me after every push, and inconveniently a few hours later.

It's actually curious. Other integrations of bots that just update an existing comment manage to not trigger a notification, e.g. this one https://github.com/flatsurf/flatsurf/pull/336#issuecomment-1901226927 (code coverage.) The first post, notifies you, the following updates are silent.

In any case, I would just add instructions to the checklist where to find the documentation. The posts by the bot here contain no information I think, other than the link to the documentation that can be found elsewhere. (That's different in the case of the coverage bot linked above. That bot alerts you that the coverage dropped for example.)

I am happy to update the PR if that seems reasonable to you @mkoeppe.

**Comment by kwankyu** [comment]
-1 (strong if you ask) 

I use the links in the comment while reviewing any PR. The doc-preview comment also contains "changes", which could be useful (yes, not always).

I am actually surprised that some people find the doc-preview comment on github interface is annoying.

**Comment by mantepse** [comment]
It's noisy because every time someone updates their branch with the newest beta or release candidate I get a blue dot. I then think I have to do something (re-review, or check a comment on my own pr), but there is nothing.

**Comment by saraedum** [comment]
> > In the case there might be a vote at some point: I am strongly in favour of removing it. I check the doc before I give Positive review, but the bot notifies me after every push, and inconveniently a few hours later.
> 
> It's actually curious. Other integrations of bots that just update an existing comment manage to not trigger a notification, e.g. this one [flatsurf/flatsurf#336 (comment)](https://github.com/flatsurf/flatsurf/pull/336#issuecomment-1901226927) (code coverage.) The first post, notifies you, the following updates are silent.
> 
> In any case, I would just add instructions to the checklist where to find the documentation. The posts by the bot here contain no information I think, other than the link to the documentation that can be found elsewhere. (That's different in the case of the coverage bot linked above. That bot alerts you that the coverage dropped for example.)

@kwankyu would that work for you?

**Comment by mkoeppe** [comment]
@saraedum Our [review instructions](https://deploy-livedoc--sagemath.netlify.app/html/en/developer/review) already say very clearly that the built doc needs to be inspected. As I said above in https://github.com/sagemath/sage/pull/37387#issuecomment-1951395012, the point is that this step has been skipped too many times in the past. 

I understand that some may find the Github notifications noisy, but to me this seems a very minor DX concern compared to the project's concern to have quality documentation for its users.

And for me as a developer with a high volume of PRs, being able to work in a notification-driven way (not waiting for a local docbuild to finish and not having to take manual note of the task to check the built doc later) is crucial for productivity.

**Comment by mkoeppe** [comment]
What should be investigated instead:
- Fixing (again) the suppression of inconsequential differences in the diff -- as can be seen here in the https://deploy-preview-37387--sagemath.netlify.app/changes, the mathjax CDN URL changed 
- Suppressing the preview comment when there is no change to the documentation.

**Comment by mkoeppe** [comment]
@kwankyu Another thought: Would we be able to download a previously deployed documentation archive from netlify? And then suppress the preview comment if there was no change compared to that?

**Comment by mkoeppe** [comment]
> Why would people skip that step if it's in the checklist?

Not sure to what depth you're hoping for me to explain human behavior...

My point is that they do.

**Comment by saraedum** [comment]
Why would they *not* skip the comment from the bot? Do you have evidence that it makes a difference?

**Comment by mantepse** [comment]
> And for me as a developer with a high volume of PRs, being able to work in a notification-driven way (not waiting for a local docbuild to finish and not having to take manual note of the task to check the built doc later) is crucial for productivity.

Isn't there an option to get notified by email?

**Comment by saraedum** [comment]
> Isn't there an option to get notified by email?

There's probably nothing builtin. I guess there must be a way (preexisting bot say) to get notified when the pipelines are finished which seems like a very common use case that people might have.

**Comment by mkoeppe** [comment]
> Why would they _not_ skip the comment from the bot? Do you have any reason to believe that that one is more effective than a checklist item?

Reminders work, in particular if they include the specific thing that needs checking (the diff links)

**Comment by mkoeppe** [comment]
FWIW, I'm getting the notifications by email. and the emails are what prompts me to check the built docs.
I don't use the github notifications thingy in the browser.

**Comment by saraedum** [comment]
> > Why would they _not_ skip the comment from the bot? Do you have any reason to believe that that one is more effective than a checklist item?
> 
> Reminders work, in particular if they include the specific thing that needs checking (the diff links)

For me, they do not work. Personally, I find it really annoying if people "remind" me about things that they asked me to do less than a week ago. In this particular case, these reminders just mean that I ignore all sagemath/sage email from GitHub (though I could fix this with a filter) and also that I cannot navigate my inbox for sagemath here on GitHub at times.

File: .github/workflows/doc-publish.yml
Comment (saraedum) [inline .github/workflows/doc-publish.yml:73]: So, if we just drop this line. Then the noise is gone as well and we get a single comment that updates (silently) instead?

Comment (saraedum) [reply .github/workflows/doc-publish.yml:73]: Would that be an acceptable approach to @kwankyu and @mkoeppe  (and everybody else of course)?

Comment (mkoeppe) [reply .github/workflows/doc-publish.yml:73]: No. Why do you ask the same question again? It has been clearly answered. The notification is what people who are doing the work want to get.

Comment (mantepse) [reply .github/workflows/doc-publish.yml:73]: @mkoeppe , actually no. The notification gets on my nerves. But more importantly, it may lead to me not noticing important messages, because I click "mark all as read"

Comment (saraedum) [reply .github/workflows/doc-publish.yml:73]: > Why do you ask the same question again?

I do by no means see this as a repeated question.

And I am trying to find some middle ground here between those who are annoyed by these notifications and those who want to keep them. With the proposed change, you still get all the relevant links. But if you want push notifications, you'd have to get these by other means.

> The notification is what people who are doing the work want to get.

I hope you are not trying to imply anything with "people who are doing the work". There seems to be a substantial amount of people here who do **not** want these notifications.

So, unless I misunderstood you here. You want a notification and email every time the workflow finishes. So you want, what others consider spam. That's your opinion and you are entitled to it. But since that's precisely what others do not want, we have apparently no hope to come to a compromise here.

**Comment by kwankyu** [comment]
Sorry for being unresponsive. Curiously I missed notifications of the recent comments on this thread.

> It's noisy because every time someone updates their branch with the newest beta or release candidate I get a blue dot. I then think I have to do something (re-review, or check a comment on my own pr), but there is nothing.

So the reason why people get annoyed is that they get a notification email for each update of the doc preview comment. That is what "a blue dot" means. Right?

I said I don't get those notification emails. But I was wrong. I am receiving the emails, but I also have this gmail filter:

Screenshot 2024-03-31 at 8 56 25 PM

which simply deletes the email.  So I got never annoyed by an email for the doc preview comment (both the initial one and the afterward updates). I visit the PR  and click the doc preview comment always on github web interface.

Do others have a difficulty in setting up such an email filter? Or such an email filter does not work well with your workflow?

**Comment by kwankyu** [comment]
> > In any case, I would just add instructions to the checklist where to find the documentation. The posts by the bot here contain no information I think, other than the link to the documentation that can be found elsewhere. (That's different in the case of the coverage bot linked above. That bot alerts you that the coverage dropped for example.)
> 
> @kwankyu would that work for you?

No. The point of the doc preview comment is that when I see that, I know that the doc preview is up to date and ready.

**Comment by kwankyu** [comment]
> What should be investigated instead:
> 
> * Fixing (again) the suppression of inconsequential differences in the diff -- as can be seen here in the https://deploy-preview-37387--sagemath.netlify.app/changes, the mathjax CDN URL changed

This would fix the broken "changes.html": https://github.com/sagemath/sage/pull/37689

**Comment by kwankyu** [comment]
It seems that what annoys people is the notification emails, not the comment. (1) If there's no way for people to avoid the notification emails, and (2) if reducing the number of updates of the comment is the agreed way to go, then we may consider 

- Suppressing the preview comment when there is no change to the documentation.

- download a previously deployed documentation archive from netlify, and then suppress the preview comment if there was no change compared to that.

But before the conditions are met, I don't think they are worth while to consider or implement.

**Comment by tornaria** [comment]
> It seems that what annoys people is the notification emails, not the comment. (1) If there's no way for people to avoid the notification emails, and (2) if reducing the number of updates of the comment is the agreed way to go, then we may consider
> 
>     * Suppressing the preview comment when there is no change to the documentation.
> 
>     * download a previously deployed documentation archive from netlify, and then suppress the preview comment if there was no change compared to that.

Not just notification *emails*, there's also the notifications in github website which appear when a comment is added to a PR one subscribes to. It really feels like spam, sometimes one gets 10-20 notifications from this which hides other, presumably more relevant, notifications (this happens particularly after a new tag when lots of branches are merged; we don't get notified that the branches were merged, and we should not be notified of this either)

Isn't there a way to add the "comment" in an "out of band" way? For example, when one changes labels, or requests a review, pushes new commits, etc, all these actions result in a "comment" added to the PR which you can see, but it won't generate a notification (neither in gh website nor by email).

> But before the conditions are met, I don't think they are worth while to consider or implement.

I would say spam should always be opt-in. So, until this can be implemented in a way that doesn't add undesired notifications, it's better to disable it.

**Comment by kwankyu** [comment]
> Not just notification _emails_, there's also the notifications in github website which appear when a comment is added to a PR one subscribes to. It really feels like spam, sometimes one gets 10-20 notifications from this which hides other, presumably more relevant, notifications (this happens particularly after a new tag when lots of branches are merged; we don't get notified that the branches were merged, and we should not be notified of this either)

You mean this

Screenshot 2024-03-31 at 11 53 57 PM

Hmm. I never bothered this blue dot. Thus I have 1270 unread notifications :-)

**Comment by kwankyu** [comment]
> Isn't there a way to add the "comment" in an "out of band" way? For example, when one changes labels, or requests a review, pushes new commits, etc, all these actions result in a "comment" added to the PR which you can see, but it won't generate a notification (neither in gh website nor by email).

So you get notifications (both in gh website and by emails) for all these actions (change labels, etc.) for a PR. That must be very annoying. I seem to receive (perhaps through the email filters) much less emails. I don't get emails for label changes.

Then even if we disable doc preview comments, you would still get tons of spam notifications. Am I right?

**Comment by tornaria** [comment]
> > Isn't there a way to add the "comment" in an "out of band" way? For example, when one changes labels, or requests a review, pushes new commits, etc, all these actions result in a "comment" added to the PR which you can see, but it won't generate a notification (neither in gh website nor by email).
> 
> So you get notifications (both in gh website and by emails) for all these actions (change labels, etc.) for a PR. That must be very annoying. I seem to receive (perhaps through the email filters) much less emails. I don't get emails for label changes.

No, that's the point. These actions leave trails in the PR, but don't send notifications. I'm just saying that maybe documentation information could be the same (no idea if the api allows for that, what I called "out of band" above).

> Then even if we disable doc preview comments, you would still get tons of spam notifications. Am I right?

Other than doc preview comments, I only get notifications for comments made by people (and also for actions like "close" -- which just happened in a bunch of PRs as 10.4.beta1 was released).

**Comment by jhpalmieri** [comment]
Is there no way to allow users to customize this, so those who want the current behavior can keep it, while those who do not want the various notifications can avoid them?

**Comment by saraedum** [comment]
> > > Isn't there a way to add the "comment" in an "out of band" way? For example, when one changes labels, or requests a review, pushes new commits, etc, all these actions result in a "comment" added to the PR which you can see, but it won't generate a notification (neither in gh website nor by email).
> > 
> > 
> > So you get notifications (both in gh website and by emails) for all these actions (change labels, etc.) for a PR. That must be very annoying. I seem to receive (perhaps through the email filters) much less emails. I don't get emails for label changes.
> 
> No, that's the point. These actions leave trails in the PR, but don't send notifications. I'm just saying that maybe documentation information could be the same (no idea if the api allows for that, what I called "out of band" above).

You cannot comment "out of band" but you can make the updates to the comment "out of band". You do not get notified when somebody modifies a comment. We can do the same thing here. Just update the documentation preview comment (setting `recreate: false`.) So, you only get the initial notification when the comment gets created but no update notification after that. (And there's no option to enable such behavior in GitHub directly, see https://github.com/isaacs/github/issues/310.) But see the discussion at https://github.com/sagemath/sage/pull/37387#discussion_r1541978770, these re-notifications are what @mkoeppe wants to actually keep.

**Comment by mkoeppe** [comment]
> we may consider
> 
> * Suppressing the preview comment when there is no change to the documentation.
> * download a previously deployed documentation archive from netlify, and then suppress the preview comment if there was no change compared to that.

I've opened #37739 for this. I consider this a wishlist item.

**Comment by tornaria** [comment]
Glad that the is a clear way forward, so this PR can be merged now.

**Review by tornaria — APPROVED** [review]

**Comment by saraedum** [comment]
@tornaria I don't read @mkoeppe's comment as him agreeing to merging this PR but probably he can clarify.

Also, I don't think we should merge this as is. From my point of view, we need to at least tell developers where they can find the documentation, e.g., in the PR checklist.

**Comment by saraedum** [comment]
I created https://github.com/sagemath/sage/pull/37740 as a solution that tries to find some middle ground as outlined above.

**Comment by tornaria** [comment]
> @tornaria I don't read @mkoeppe's comment as him agreeing to merging this PR but probably he can clarify.

The PR description has two :+1: (not counting the author) and one :-1: if I'm not mistaken. But that was not my point.

> Also, I don't think we should merge this as is. From my point of view, we need to at least tell developers where they can find the documentation, e.g., in the PR checklist.

I thought this was addressed here:
https://github.com/sagemath/sage/pull/37387#issuecomment-1962924245

**Comment by mkoeppe** [comment]
> @tornaria I don't read @mkoeppe's comment as him agreeing to merging this PR but probably he can clarify.

@saraedum That's right.

**Comment by mkoeppe** [comment]
> > we may consider
> > 
> > * Suppressing the preview comment when there is no change to the documentation.
> > * download a previously deployed documentation archive from netlify, and then suppress the preview comment if there was no change compared to that.
> 
> I've opened #37739 for this. I consider this a wishlist item.

Apparently I need to be more direct.

In https://github.com/sagemath/sage/pull/37387#issuecomment-2023716887, https://github.com/sagemath/sage/pull/37387#issuecomment-2028718420, now #37739, we have sketched a clear approach that would reduce the notifications in a way that would *not* diminish the value of this productivity tool. This should be given proper consideration; so far there has not been any response.

**Comment by saraedum** [comment]
> > Also, I don't think we should merge this as is. From my point of view, we need to at least tell developers where they can find the documentation, e.g., in the PR checklist.
> 
> I thought this was addressed here: [#37387 (comment)](https://github.com/sagemath/sage/pull/37387#issuecomment-1962924245)

True. As an alternative, we could just consider to update the PR template to tell people where to find this. If we don't tell people, I don't think anybody would click there :)

**Comment by saraedum** [comment]
> What should be investigated instead:
> 
>     * Fixing (again) the suppression of inconsequential differences in the diff -- as can be seen here in the https://deploy-preview-37387--sagemath.netlify.app/changes, the mathjax CDN URL changed

I don't see how this is immediately related to the problem that we are trying to solve here. (Sure, it makes one of the links somewhat less useful but this PR is about the notifications that people get.)

>     * Suppressing the preview comment when there is no change to the documentation.

That's an alternative approach but it's substantially more work to implement. And I'd rather not receive notifications at all to be honest.

**Comment by mkoeppe** [comment]
> > * Suppressing the preview comment when there is no change to the documentation.
> 
> That's an alternative approach but it's substantially more work to implement. 
>
> And I'd rather not receive notifications at all to be honest.

That doesn't rise to the standard "proper consideration".

**Comment by mkoeppe** [comment]
@saraedum In https://github.com/sagemath/sage/pull/37387#issuecomment-2035707014 I wrote that the solution sketched in https://github.com/sagemath/sage/issues/37739 "should be given proper consideration". Saying that it's "substantially more work to implement" and mentioning your previously stated personal preference regarding the notifications is not "proper consideration". I'm of course aware that it's more work to implement; that's why I've marked it as a "wishlist item". I would suggest that those who feel bothered by the notifications invest an appropriate amount of effort in it.

**Comment by saraedum** [comment]
> @saraedum In [#37387 (comment)](https://github.com/sagemath/sage/pull/37387#issuecomment-2035707014) I wrote that the solution sketched in #37739 "should be given proper consideration". Saying that it's "substantially more work to implement" and mentioning your previously stated personal preference regarding the notifications is not "proper consideration".

I disagree. I consider that I have given it proper consideration.

> I'm of course aware that it's more work to implement; that's why I've marked it as a "wishlist item". I would suggest that those who feel bothered by the notifications invest an appropriate amount of effort in it.

I also don't agree here. Why should those who would like to completely get rid of the notification invest into making it better? That's why I propose to not send further notifications after the first one. It's a middle ground. (That it's easy to implement is another bonus of course.)

Let's not get caught up on the technicalities of who should be doing what supposedly. (Just worried that we're going off topic here. I am happy to chat about this on Zulip or elsewhere.)

**Comment by saraedum** [comment]
> > Why should those who would like to completely get rid of the notification invest into making it better?
> 
> Because that's how consideration and collaboration should work in our community.

I don't see it that way. And I don't think it's realistic that people will want to spend their free time like that.

> The purpose and value of the documentation preview notification feature have been explained clearly and thoroughly.

Indeed. But I think it's also been made clear why several people find this feature very annoying.

> I have seen nothing brought forward here that would explain why piling up notifications should be considered more than a minor developer experience concern, and in particular why that should weigh more than what the feature offers.

I don't think it's considered "minor" by those who don't like it. Certainly, I consider it major to the workflow you describe but obviously that's colored by me not using that workflow. Also, we have seen several people say that they are annoyed by this feature but so far only you have said that you use it for your workflow. (Am I missing somebody?)

> > That's why I propose to not send further notifications after the first one. It's a middle ground.
> 
> The standard warning about the fallacy of the golden mean applies. This proposal is not suitable as a compromise

I am sorry, but from my perspective neither is your insistence on keeping the notifications essentially as they are (I understand that you see #37739 as a compromise of sorts but I don't see this as making a change to the actual problem here.)

Again, I don't think this ticket is the right place to discuss who should work on what and who should decide on what.

**Comment by mkoeppe** [comment]
> > > That's why I propose to not send further notifications after the first one. It's a middle ground.
> > 
> > The standard warning about the fallacy of the golden mean applies. This proposal is not suitable as a compromise
> 
> I am sorry, but from my perspective neither is your insistence

Julian, kind reminder that this word has no place in a discussion.

> on keeping the notifications essentially as they are (I understand that you see #37739 as a compromise of sorts but I don't see this as making a change to the actual problem here.)

How can sending fewer notifications (per #37739) not be relevant as a possible compromise?

**Comment by tornaria** [comment]
> > I am sorry, but from my perspective neither is your insistence
> 
> Julian, kind reminder that this word has no place in a discussion.

@saraedum the technical word is:

From WordNet (r) 3.0 (2006) [wn]:

  intransigence
      n 1: the trait of being intransigent; stubbornly refusing to
           compromise [syn: {intransigency}, {intransigence}]

  intransigent
      adj 1: impervious to pleas, persuasion, requests, reason; "he is
             adamant in his refusal to change his mind"; "Cynthia was
             inexorable; she would have none of him"- W.Churchill; "an
             intransigent conservative opposed to every liberal
             tendency" [syn: {adamant}, {adamantine}, {inexorable},
             {intransigent}]

**Comment by mkoeppe** [comment]
Another possible variant: Have only the PR author receive the docbuild notification.
- I believe that the notifications when a workflow fails are sent out only to the PR author.
- So we can just have the `doc-publish.yml` workflow report a *failure*.
- How to educate users about it: "It's a failure because there is a change in the formatted HTML documentation that has not been reviewed yet." (So, ideally, there should be a way to mark it as reviewed.)

**Comment by jhpalmieri** [comment]
> > > I am sorry, but from my perspective neither is your insistence
> > 
> > 
> > Julian, kind reminder that this word has no place in a discussion.
> 
> @saraedum the technical word is:
> 
> From WordNet (r) 3.0 (2006) [wn]:
> 
> intransigence n 1: the trait of being intransigent; stubbornly refusing to compromise [syn: {intransigency}, {intransigence}]
> 
> intransigent adj 1: impervious to pleas, persuasion, requests, reason; "he is adamant in his refusal to change his mind"; "Cynthia was inexorable; she would have none of him"- W.Churchill; "an intransigent conservative opposed to every liberal tendency" [syn: {adamant}, {adamantine}, {inexorable}, {intransigent}]

Gonzalo, you have stated this point of view publicly already, and Matthias objected to your previous post. There is no need to repeat it.

**Comment by mantepse** [comment]
> Another possible variant: Have only the PR author receive the docbuild notification.

It would make more sense to me if the reviewers are notified.

**Comment by mkoeppe** [comment]
@mantepse Let me try to explain the rationale for my suggestion better: In the notification-based workflow, a developer is working on multiple PRs simultaneously. After pushing a change to one PR, they can immediately switch to working on a different PR. When notifications come in (either "CI failed" or "documentation ready for inspection"), they can act on the notifications -- either by fixing the failure or by checking the documentation. In this way, the notifications enable a work schedule that is free of waiting time. 
I agree with you that reviewers may also benefit from the documentation notification as a reminder; but I'm not aware of a reviewing mode that benefits as clearly from the scheduling aspect of this mechanism.

**Comment by mantepse** [comment]
Ok. Well, I usually work on a single topic, and my efficiency is limited not by waiting time but rather my brain and interest. So, this workflow is not for me.

**Comment by mkoeppe** [comment]
This workflow also works as well for developing several related PRs on the same topic.

**Comment by tobiasdiez** [comment]
Closing this in favor of the compromise https://github.com/sagemath/sage/issues/37739.

**Comment by jhpalmieri** [comment]
I'm removing the "disputed" tag since this was closed by the author.

**Comment by jhpalmieri** [comment]
> > I am sorry, but from my perspective neither is your insistence
> 
> Julian, kind reminder that this word has no place in a discussion.

On behalf of the Code of Conduct Committee: we do not see any problem with the word "insistence". It does not have the same negative connotations as some perhaps similar words. It is also not clear what "reminder" refers to, since the word has not come up before this.

**Pull request closed without merging**
